text/0052-reserved-namespaces.md

```diff
@@ -0,0 +1,108 @@
+# Reserve Identifiers for Internal Use
+
+## Related issues and PRs
+
+- Reference Issues: (fill in existing related issues, if any)
+- Implementation PR(s): (leave this empty)
+
+## Timeline
+
+- Started: 2024-02-14
+
+## Summary
+
+This RFC extends the reservation of the `__cedar` namespace in schema in [RFC24](https://github.com/cedar-policy/rfcs/blob/main/text/0024-schema-syntax.md) and
+proposes reserving all namespaces that start with two underscores (`__`),
+forbidding their use in Cedar policies, schema, and entity data so that they can be used by future Cedar language features.
+
+## Basic example
+
+Under this RFC the names `__cedar::Long`, `__Long`, and `__namespace::User` will all become invalid in all contexts.
+
+These would be rejected in a policy
+
+```cedar
+permit(principal == __cedar::User::"alice", actions, resource);
+```
+
+in a schema
+
+```
+namespace __DocCloud {
+  ...
+}
+```
+
+and in entity data
+
+```json
+[
+    {
+        "uid": { "__entity": {
+            "type": "__cedar::User",
+            "id": "alice"
+        }},
+        "attrs": {},
+        "parents": []
+    }
+]
+```
+
+In this entity data example, the `__entity` JSON key does not demonstrate the new reserved namespace mechanism.
+It is a place were we already assign special significance to some identifiers starting with `__` in entity data.
+
+## Motivation
+
+In designing [RFC24](https://github.com/cedar-policy/rfcs/blob/main/text/0024-schema-syntax.md) we found that we needed to disambiguate user defined identifiers from built-in primitive and extension types.
+We accomplished this by reserving only the `__cedar` namespace inside of schema,
+but we anticipate that we may need more reserved namespaces in more contexts in the future.
```

Comment:
I'm not totally convinced by this motivation.  If we have `__cedar` as a reserved namespace, what would cause us to need other `__` namespaces?  If we do need other namespaces, we can use sub-namespaces of `__cedar` like  `__cedar::A`, `__cedar::B`, etc, which are already reserved as part of the `__cedar` namespace.